### DISCUSSION TITLE
Storybook build fail in my project after 8.2.1 release

### DISCUSSION THREAD

**Discussion by jbanovs** [storybookjs/storybook#28533] [Help] [answered]
### Summary

After updating package.lock file in my project I got an error
```
info => Using default Webpack5 setup
=> Failed to build the preview
SB_BUILDER-WEBPACK5_0002 (WebpackInvocationError): Module not found: Error: Can't resolve 'history' in './node_modules/.pnpm/@storybook+core@8.2.1/node_modules/@storybook/core/dist/router'
    at ./node_modules/.pnpm/@storybook+builder-webpack5@8.2.1_esbuild@0.21.5_storybook@8.2.1_typescript@5.0.4_webpack-cli@5.1.4/node_modules/@storybook/builder-webpack5/dist/index.js:1:8835
    at ./node_modules/.pnpm/webpack@5.92.1_esbuild@0.21.5_webpack-cli@5.1.4/node_modules/webpack/lib/HookWebpackError.js:68:3
    at _done (eval at create (./node_modules/.pnpm/tapable@2.2.1/node_modules/tapable/lib/HookCodeFactory.js:33:10), <anonymous>:14:1)
    at eval (eval at create (./node_modules/.pnpm/tapable@2.2.1/node_modules/tapable/lib/HookCodeFactory.js:33:10), <anonymous>:44:22)
```

It is possible to fix the issue adding react-router-dom@6.3 (<6.4) in my project.

Also I've tried to override all storybook dependencies back to 8.1.11 and it works.

I think the problem is that my build uses cjs files and react-router-dom is not in direct dependencies of storybook.

There are `define(["exports", "react", "history"]` in [@storybook/core@8.2.1/router](https://unpkg.com/@storybook/core@8.2.1/dist/router/index.cjs) but there aren't in [@storybook/router@8.1.11](https://unpkg.com/@storybook/router@8.1.11/dist/index.js)

I wouldn't want to add extra dependencies in my package or pin versions of storybook. What the better way to fix my issue?

**Comment by jbanovs** [comment]
Also I faced another issue trying to run tests
```
 web/stories/list.stories.test.tsx
  ● Test suite failed to run

    Cannot find module 'storybook/internal/preview-api' from 'node_modules/.pnpm/@storybook+react@8.2.1_react-dom@18.3.1_react@18.3.1_storybook@8.2.1_typescript@5.0.4/node_modules/@storybook/react/dist/index.js'
```

**Reply by ndelangen** [reply]
Could you please supply a reproduction repository?

**Comment by ndelangen** [comment]
I'm confused as to why your storybook preview builder is doing anything with the router package.
That package is for the manager only, the preview should never need/use it, AFAIK.

This code here (the CJS version!), [does indeed contain](https://unpkg.com/@storybook/core@8.2.1/dist/router/index.cjs):
```js
// ../node_modules/react-router/umd/react-router.production.min.js
var Wa = Q((bt, Ba) => {
  (function(t, e) {
    typeof bt == "object" && typeof Ba < "u" ? e(bt, require("react"), Ye()) : typeof define == "function" && define.amd ? define(["exports",
    "react", "history"], e) : e((t = t || self).ReactRouter = {}, t.React, t.HistoryLibrary);
  })(bt, function(t, e, a) {
```
So it's using the UMD version of `react-router`.?
I wonder why?

I found that there IS a version of `history` embedded:
```js
// ../node_modules/history/umd/history.development.js
var Da = Q((vt, qa) => {
  (function(t, e) {
    typeof vt == "object" && typeof qa < "u" ? e(vt) : typeof define == "function" && define.amd ? define(["exports"], e) : (t = typeof globalThis <
    "u" ? globalThis : t || self, e(t.HistoryLibrary = {}));
  })(vt, function(t) {
    "use strict";
    function e() {
```

But why is your webpack bundling that at all? And the CJS file?!

I can look into that.

Regarding this:
```
Cannot find module 'storybook/internal/preview-api' from 'node_modules/.pnpm/@storybook+react@8.2.1_react-dom@18.3.1_react@18.3.1_storybook@8.2.1_typescript@5.0.4/node_modules/@storybook/react/dist/index.js'
```
I think you may need to upgrade your test runner to support export maps :(

**Reply by jbanovs** [reply, marked as answer]
Thank you for comments!

I realised why preview uses manager package. I have a decorator that sends messages from preview to my addon.
```typescript
  const channel = useMemo(() => {
    const ch = createBrowserChannel({ page: 'preview' });
    addons.setChannel(ch);
    return ch;
  }, []);
```
But I used wrong package to import addons 🤦🏻‍♂️
```typescript
import { addons } from '@storybook/manager-api';
```
There definitely must be `@storybook/preview-api`. May be it is possible to check it somehow?

And, yes, I've managed to fix the issue with unit tests upgrading to latest version of `jest`. May be it is better to highlight this in the migration guide or change log?
